### PR TITLE
Add SignerEvent::NewNakamotoBlock and do not update a block to GloballyAccepted until event arrives

### PR DESCRIPTION
**Pull request by jferrant**
Closes: https://github.com/stacks-network/stacks-core/issues/5514

### CODE REVIEW

**Review by hstove — CHANGES_REQUESTED** [review]
I think an integration test would be needed, otherwise LGTM!

Is `/new_block` always sent to an observer, regardless of the event keys they've set? It seems so, but I'd want to be 100% on that

**Comment by jferrant** [comment]
> I think an integration test would be needed, otherwise LGTM!
> 
> Is `/new_block` always sent to an observer, regardless of the event keys they've set? It seems so, but I'd want to be 100% on that

It is always yes. But I cannot for the life of me test this without so much jankiness that I think it defeats the purpose. We already have multiple tests that pass through this exact code so I think its fine to not have an integration test (essentially order of ops is:
1. Locally rejected for reasons. 
2. We see 70% acceptance. Mark locally accepted.
3. Get new block event: mark globally accepted".

If 2 and 3 happen out of order, it doesn't matter. we are fine.

If we forcibly cause the signer to ignore both signatures and broadcasted events but still have 70% of signers approve it...we still are fine again because when we go to verify against the node...it returns "Invalid chain length"  Basically I cannot force this situation to occur. I have tried preventing brodcasting of new burn blocks. I have also tried paysing miner accepting signatures. I have also forced signers to ignore other signers signatures. This MAY work if I can manage to simulate a block at the same height that passes all other validation checks. But again...does this actually REALLY test what we awnt to test?

**Review by hstove — DISMISSED** [review]
LGTM!

File: stacks-signer/src/v0/signer.rs
Comment (obycode) [inline stacks-signer/src/v0/signer.rs:69]: You can use `TestFlag` for this.

File: stacks-signer/src/v0/signer.rs
Comment (obycode) [inline stacks-signer/src/v0/signer.rs:408]: I see why you did this, but it looks a little wrong. Would this be more clear?
```rust
        #[cfg(any(test, feature = "testing"))]
        let mut block_info = BlockInfo::from(block_proposal.clone());
        #[cfg(not(any(test, feature = "testing")))]
        let block_info = BlockInfo::from(block_proposal.clone());
```
I don't have a strong opinion on this, just a suggestion.

File: stacks-signer/src/v0/signer.rs
Comment (obycode) [inline stacks-signer/src/v0/signer.rs:452]: I'm not clear on why this is no longer needed. Can you explain that part?

File: testnet/stacks-node/src/event_dispatcher.rs
Comment (obycode) [inline testnet/stacks-node/src/event_dispatcher.rs:74]: `TestFlag`

**Comment by jferrant** [comment]
I am going to rewrite the integration test to do the following as I think this is the only way I can verifiably trigger the case that I want:

Miner A mines a regular tenure A.
Force miner to ignore signatures.
Skip block broadcast for Signers.
Force one signer to reject any incoming blocks
Force miner A to mine a block via transfer tx. (block N)
Signers will eventually ALL mark the block as locally accepted (70% threshold will approve so even rejecting signer will remark the block as locally accepted)
Allow all signers to accept/reject at will incoming blocks.
Reenable block broadcast.
Reenable miner to listen to signatures.
Mine another BTC block so Miner A wins a new tenure B.
Miner A should attempt to mine a block N'.
This block should be accepted by all signers as reorging LOCALLY accepted blocks across tenures is acceptable.
This block N' will be marked as globally accepted. 
Propose block N''. This will be rejected outright by the signers with a sortitionviewmismatch (won't even submit for block proposal approval)

File: stacks-signer/src/v0/signer.rs
Comment (jferrant) [reply stacks-signer/src/v0/signer.rs:408]: Sure :)

File: stacks-signer/src/v0/signer.rs
Comment (jferrant) [reply stacks-signer/src/v0/signer.rs:452]: We do not store the block in the database. Why bother mutating it to never store it?

Comment (obycode) [reply stacks-signer/src/v0/signer.rs:452]: Ah, right. Thanks!

**Review by obycode — APPROVED** [review]
LGTM!

File: libsigner/src/events.rs
Comment (jcnelson) [inline libsigner/src/events.rs:556]: Instead of making a bespoke and scoped struct, it might be cleaner to just add a `fn from_slice(bytes: &[u8]) -> Result<SignerEvent<T>, EventError>` function to `SignerEvent<T>`.  Then, you could put all of the decode logic into that function, and just pass in `body.as_bytes()`.

File: stacks-common/src/util/mod.rs
Comment (jcnelson) [inline stacks-common/src/util/mod.rs:38]: Because this is all test code, can this be placed into a designated test file or module?

File: stacks-signer/src/v0/signer.rs
Comment (jcnelson) [inline stacks-signer/src/v0/signer.rs:1166]: Same idea here -- can this go into a dedicated test module?

**Review by jcnelson — COMMENTED** [review]
This looks great.  Thank you for addressing this!  Just a few minor comments before approving.

I like what you've done to clean up the test flags.  Do you think you could take a moment and add a couple sentences on what each one does (or open an issue for doing so if it's too out-of-scope), so future test authors know how to use them?  Thanks!

File: stacks-signer/src/v0/signer.rs
Comment (jferrant) [reply stacks-signer/src/v0/signer.rs:1166]: Can do :D

File: stacks-common/src/util/mod.rs
Comment (jferrant) [reply stacks-common/src/util/mod.rs:38]: Can do :D

File: libsigner/src/events.rs
Comment (jferrant) [reply libsigner/src/events.rs:556]: I don't think I fully understand what you mean by passing in body.as_bytes(). But I did make a more generic process_event function that will correctly deserialize the request bytes so long as the event implements serde and try_from for SignerEvent<T>.

**Review by obycode — APPROVED** [review]

**Review by obycode — APPROVED** [review]

File: testnet/stacks-node/src/event_dispatcher.rs
Comment (hstove) [inline testnet/stacks-node/src/event_dispatcher.rs:80]: Nit: I think we want to move to std::sync::LazyLock / LazyCell https://github.com/stacks-network/stacks-core/issues/5035

File: stacks-common/src/util/tests.rs
Comment (hstove) [inline stacks-common/src/util/tests.rs:48]: Love that you made this generic!

Comment (hstove) [reply stacks-common/src/util/tests.rs:48]: And that you moved it to stacks-common 😀

**Review by hstove — APPROVED** [review]
One tiny nit about lazy_static. The test-related refactors are great!

**Review by jcnelson — APPROVED** [review]

File: testnet/stacks-node/src/event_dispatcher.rs
Comment (jferrant) [reply testnet/stacks-node/src/event_dispatcher.rs:80]: I will leave it for now but will keep in mind for a follow on. Just don't want to hold this up any further XD

**Pull request merged**
